t/03-testapi.t

```diff
@@ -683,4 +683,19 @@ subtest 'autoinst_url' => sub {
     is(autoinst_url('foo'), 'http://localhost:1/foo', 'we can configure the hostname that autoinst_url returns');
 };
 
+subtest '_calculate_clickpoint' => sub {
+    my $mock_testapi = Test::MockModule->new('testapi');
+    my @areas        = ({x => 10, y => 10, w => 20, h => 30});
+    $mock_testapi->redefine(assert_screen => {area => \@areas});
+
+    my %fake_needle_area = {x => 100, y => 100, w => 50, h => 40};
+    my %fake_click_point = {xpos => 20, y => 10};
+
+    $cmds = [];
+    # Everything is provided.
+    my ($x, $y) = testapi::_calculate_clickpoint('foo', %fake_needle_area, %fake_click_point));
```

Comment:
Looks like there's one `)` too much. Besides, you should likely pass the hashes as reference, e.g. `… \%fake_needle_area, \%fake_click_point …`.

Reply:
What is the difference between this suggestion and the original one? What does "passing as reference" means? Sorry, told you I am not quite experienced in Perl :D

Reply:
in perl, when e.g. calling a subroutine with an arg that's an array or hash, you can either send *the actual array or hash itself*, or you can send a *reference* to it. A reference is a scalar which refers to a hash or an array (or other things - you can have a coderef, for instance, a reference to a block of code).

You usually don't have to worry about this in Python because Python takes care of it for you and doesn't give you any choices.

https://perldoc.perl.org/perlreftut.html is a good starting point on this stuff. The example @Martchus gave is the way to create a reference to an existing hash: if you have a hash `%foo`, then `\%foo` creates a reference to that hash. You can pass an anonymous reference when calling a subroutine - e.g. `mysub(\%foo);` calls `mysub` with an anonymous reference to `%foo` as the arg - or you can assign it to a scalar: `my $fooref = \%foo;` and then do stuff with the scalar, e.g. `mysub($foo);`.

If you want a subroutine to take arrays or hashes as args it's almost always best to use references, because passing actual arrays and hashes as subroutine args doesn't really work the way you might expect. Especially if the sub takes more than one arg.

Reply:
if you look at `_calculate_checkpoint` itself, you can see it's expecting to be passed scalars:

    sub _calculate_clickpoint {
        my ($needle_to_use, $needle_area, $click_point) = @_;

and the rest of the code actually shows that `$needle_area` and `$click_point` are (mostly) expected to be hash refs, once you understand references.

    if (!$click_point || $click_point eq 'center') {
        $click_point = {
            xpos => $needle_area->{w} / 2,
            ypos => $needle_area->{h} / 2,
        };
    }

First off, see that use of `->`? That is the convenient way to get a value out of a hash *reference*. It's not how you'd get the value out of a hash. If you had a *hash* `%needle_area` you'd use `$needle_area{w}`.

You can also see that here we're handling the case where `$click_point` is not passed at all or is passed not as a hash ref but as the special string value 'center', and what we're actually doing in those cases is *redefining `$click_point` as a hash ref*. Note we use curly braces to define `$click_point` here: that is a way to create a hash ref directly, as opposed to creating a named hash then creating a reference to it. If you were just creating a hash you'd do `%click_point = (<stuff>)` - you use *regular* braces. Doing `$click_point = {<stuff>}`, with curly braces, creates a new hash *ref*.

Reply:
Thank you very much, Adam, for this explanation.